### PR TITLE
Strengthen trait ordering rules

### LINKED ISSUE #10907: Requires<> is bogus.
As my recent modding experiments uncovered it, `Requires<>` turned out to be bogus. If the required trait has multiple instances on the actual actor and the trait which is required is placed between them, `Requires<>` will not force the followup instances to be loaded before the trait which requires them.

Testcase in bleed TS: add `AutoTarget` between the two `Attack*` traits to the Tick Tank (`ttnk`). The game won't crash. Add it to the end of the unit and the game will crash due to `AutoTarget` is still incapable of handling more `Attack*` traits. `AutoTarget` requires `AttackBase`.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes #10907.

Currently, if you have multiple dependencies for a trait, as long as one of them is created first then the ordering is satisfied. This means the others might be created after the thing that requires them!

Now, we strengthen the rules to require all of the dependencies to be created up front. This means the dependent can now guarantee all its dependencies are ready, rather than just at least one.

I have adjusted the tests to verify this.

This created some circular dependency chains on `UpgradeManager` in the default mods which I have resolved by doing the trait lookups after the actor is created. Without this RA and D2k fail to load.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 1]
```diff
@@ -67,20 +67,27 @@ class UpgradeState
...
+				hasUpgradeables = true;
+				foreach (var t in up.UpgradeTypes)
+					upgrades.GetOrAdd(t).Traits.Add(up);
+			}
 
-				return ret;
-			});
+			if (!hasUpgradeables)
+				throw new Exception(
+					"There are no upgrades to be managed for actor '{0}'. You are either missing some upgradeable traits, or this UpgradeManager trait is not required.".F(
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Could we check for this via `IRulesetLoaded` in the trait info?

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 2]
```diff
@@ -100,7 +100,7 @@ public IEnumerable<ITraitInfo> TraitsInConstructOrder()
 			var unresolved = source.Except(resolved);
 
 			var testResolve = new Func<Type, Type, bool>((a, b) => a == b || a.IsAssignableFrom(b));
-			var more = unresolved.Where(u => u.Dependencies.All(d => resolved.Exists(r => testResolve(d, r.Type))));
+			var more = unresolved.Where(u => u.Dependencies.All(d => resolved.Exists(r => testResolve(d, r.Type)) && !unresolved.Any(u1 => testResolve(d, u1.Type))));
```
Comment (pchote) [inline OpenRA.Game/GameRules/ActorInfo.cs]: Vagely related scope creep (not a blocker): Using LINQ here might be compact, but it makes what is otherwise a simple algorithm really difficult to parse.  How much would I have to bribe to you rewrite this block of 6 lines in a clearer way or to add comments?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 3]
```diff
@@ -17,12 +17,19 @@
...
 {
 	[Desc("Attach this to a unit to enable dynamic upgrades by warheads, experience, crates, support powers, etc.")]
-	public class UpgradeManagerInfo : ITraitInfo, Requires<IUpgradableInfo>
+	public class UpgradeManagerInfo : ITraitInfo, IRulesetLoaded
 	{
 		public object Create(ActorInitializer init) { return new UpgradeManager(init); }
+		public void RulesetLoaded(Ruleset rules, ActorInfo info)
+		{
+			if (!info.Name.StartsWith("^") && !info.TraitInfos<IUpgradableInfo>().Any())
+				throw new Exception(
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: I believe the convention is to use `YamlException` here (?)

**Comment by pchote** [comment]
This looks correct to me, but I haven't tested it so :+0.5: for now.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs (hunk 3 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Yes, please do that.

Comment (RoosterDragon) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Shall I make the same change in `TraitsInConstructOrder`? (Currently also throws Exception)

**Comment by RoosterDragon** [comment]
Actually this may cause issues for `Cloak` and `TimedUpgradeBar` which use the upgrade manager in their created callbacks. There's no guarantee that will work since the upgrade manager may not have run its creation callback yet,

I will have to think that over some more.

**Comment by RoosterDragon** [comment]
As pointed out to me on IRC, Cloak can be fixed with a frame end task and TimedUpgradeBar has a requires so that solves that.

I have tweaked the exception types and added some extra comments.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 4]
```diff
@@ -67,20 +74,17 @@ class UpgradeState
...
-			{
-				var ret = new Dictionary<string, UpgradeState>();
-				foreach (var up in init.Self.TraitsImplementing<IUpgradable>())
-					foreach (var t in up.UpgradeTypes)
-						ret.GetOrAdd(t).Traits.Add(up);
+		public UpgradeManager(ActorInitializer init) { }
 
-				return ret;
-			});
+		public void Created(Actor self)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Since you are introducing this here and don't need to call this yourself, please explicitly implement this.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 5]
```diff
@@ -67,20 +74,17 @@ class UpgradeState
 		}
 
 		readonly List<TimedUpgrade> timedUpgrades = new List<TimedUpgrade>();
-		readonly Lazy<Dictionary<string, UpgradeState>> upgrades;
 		readonly Dictionary<IUpgradable, int> levels = new Dictionary<IUpgradable, int>();
+		Dictionary<string, UpgradeState> upgrades;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: Why not `readonly Dictionary<string, UpgradeState> upgrades = new Dictionary<string, UpgradeState>();`?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs [hunk 6]
```diff
@@ -67,20 +74,17 @@ class UpgradeState
...
 
-		public UpgradeManager(ActorInitializer init)
-		{
-			upgrades = Exts.Lazy(() =>
-			{
-				var ret = new Dictionary<string, UpgradeState>();
-				foreach (var up in init.Self.TraitsImplementing<IUpgradable>())
-					foreach (var t in up.UpgradeTypes)
-						ret.GetOrAdd(t).Traits.Add(up);
+		public UpgradeManager(ActorInitializer init) { }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: You could make `UpgradeManagerInfo : TraitInfo<UpgradeManager>, IRulesetLoaded` and then remove this empty ctor definition.

File: OpenRA.Test/OpenRA.Game/ActorInfoTest.cs [hunk 7]
```diff
@@ -28,89 +27,55 @@ class MockDInfo : MockTraitInfo, Requires<MockEInfo> { }
...
-		ActorInfo CreateActorInfoFromYaml(string name, string mapYaml, params string[] yamls)
-		{
-			var nodes = mapYaml == null ? new List<MiniYamlNode>() : MiniYaml.FromString(mapYaml);
-			var sources = yamls.ToList();
-			if (mapYaml != null)
-				sources.Add(mapYaml);
+			var count = (
+				new Regex("MockD").Matches(ex.Message).Count +
+				new Regex("MockE").Matches(ex.Message).Count +
+				new Regex("MockF").Matches(ex.Message).Count) / 3.0;
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ActorInfoTest.cs]: Why not just compare the counts directly?

**Comment by pchote** [comment]
Looks good otherwise :+1:

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs (hunk 5 above)
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs]: I wanted this to throw if the methods are accessed before `Created` was called. I've added an explicit exception with a message since this will be clearer in the code and for consumers.

File: OpenRA.Test/OpenRA.Game/ActorInfoTest.cs (hunk 7 above)
Comment (RoosterDragon) [inline OpenRA.Test/OpenRA.Game/ActorInfoTest.cs]: Just didn't refactor it quite that much. I've changed the asserts.

**Comment by RoosterDragon** [comment]
Made a few minor but non-trivial changes, so removing the +1.

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 8]
```diff
@@ -100,9 +100,15 @@ public IEnumerable<ITraitInfo> TraitsInConstructOrder()
...
+			// This query detects which unresolved traits can be immediately resolved as all their direct dependencies are met.
+			var more = unresolved.Where(u =>
+				u.Dependencies.All(d => // To be resolvable, all dependencies must be satisfied according to the following conditions:
+					resolved.Exists(r => testResolve(d, r.Type)) && // There must exist a resolved trait that meets the dependency.
+					!unresolved.Any(u1 => testResolve(d, u1.Type)))); // All matching traits that meet this dependency must be resolved first.
+
+			// Continue resolving traits as long as possible.
+			// Each time we resolve some traits, this means dependencies for other traits may then be possible to satisfy in the next pass.
 			while (more.Any())
 				resolved.AddRange(more);
```
Comment (obrakmann) [inline OpenRA.Game/GameRules/ActorInfo.cs]: Why isn't this an infinite loop? `more` isn't being modified in the loop, afaics, so why should the result from `more.Any()` ever change?

Comment (pchote) [inline OpenRA.Game/GameRules/ActorInfo.cs]: `more` is an enumerable rather than a discrete collection.

**Comment by obrakmann** [comment]
I guess this is fine then, :+1:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9608ab5fa8174518d46d0a07e10baa3d8fd333f7...7360a1b115e14cc5c11ad854125718061f8ff2a5)
